pkg/gadgets/trace/dns/tracer/tracer.go

```diff
@@ -108,254 +78,133 @@ var pktTypeNames = []string{
-	57:    "RKEY",
-	58:    "TALINK",
-	59:    "CDS",
-	60:    "CDNSKEY",
-	61:    "OPENPGPKEY",
-	62:    "CSYNC",
-	63:    "ZONEMD",
-	64:    "SVCB",
-	65:    "HTTPS",
-	99:    "SPF",
-	100:   "UINFO",
-	101:   "UID",
-	102:   "GID",
-	103:   "UNSPEC",
-	104:   "NID",
-	105:   "L32",
-	106:   "L64",
-	107:   "LP",
-	108:   "EUI48",
-	109:   "EUI64",
-	249:   "TKEY",
-	250:   "TSIG",
-	251:   "IXFR",
-	252:   "AXFR",
-	253:   "MAILB",
-	254:   "MAILA",
-	255:   "*",
-	256:   "URI",
-	257:   "CAA",
-	258:   "AVC",
-	259:   "DOA",
-	260:   "AMTRELAY",
-	32768: "TA",
-	32769: "DLV",
+func pktTypeToString(pktType uint8) string {
+	pktTypeUint := uint(pktType)
+	if pktTypeUint < uint(len(pktTypeNames)) {
+		return pktTypeNames[pktType]
+	}
+
+	return "UNKNOWN"
 }
 
-const MaxDNSName = int(unsafe.Sizeof(dnsEventT{}.Name))
-
-// DNS header RCODE (response code) field.
-// https://datatracker.ietf.org/doc/rfc1035#section-4.1.1
-var rCodeNames = map[uint8]string{
-	0: "NoError",
-	1: "FormErr",
-	2: "ServFail",
-	3: "NXDomain",
-	4: "NotImp",
-	5: "Refused",
+// --- Registry changes
+
+func (g *GadgetDesc) NewInstance() (gadgets.Gadget, error) {
+	return &Tracer{}, nil
 }
 
-// parseLabelSequence parses a label sequence into a string with dots.
-// See https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.2
-func parseLabelSequence(sample []byte) (ret string) {
-	sampleBounded := make([]byte, MaxDNSName)
-	copy(sampleBounded, sample)
+func (t *Tracer) Init(gadgetCtx gadgets.GadgetContext) error {
+	if err := t.install(); err != nil {
+		t.Close()
+		return fmt.Errorf("installing tracer: %w", err)
+	}
+
+	t.ctx, t.cancel = gadgetcontext.WithTimeoutOrCancel(gadgetCtx.Context(), gadgetCtx.Timeout())
 
-	for i := 0; i < MaxDNSName; i++ {
-		length := int(sampleBounded[i])
-		if length == 0 {
-			break
-		}
-		if i+1+length < MaxDNSName {
-			ret += string(sampleBounded[i+1:i+1+length]) + "."
-		}
-		i += length
+	return nil
+}
+
+func (t *Tracer) install() error {
+	networkTracer, err := networktracer.NewTracer[types.Event]()
+	if err != nil {
+		return fmt.Errorf("creating network tracer: %w", err)
 	}
-	return ret
+	t.Tracer = networkTracer
+	return nil
 }
 
-func bpfEventToDNSEvent(bpfEvent *dnsEventTAbbrev, answers []byte, netns uint64) (*types.Event, error) {
-	event := types.Event{
+func (t *Tracer) parseDNSPacket(rawSample []byte, netns uint64) (*types.Event, error) {
+	// The sample received is a concatenation of the dnsEventT structure and the packet bytes.
+	bpfEvent := (*dnsEventT)(unsafe.Pointer(&rawSample[0]))
+	packetBytes := rawSample[unsafe.Sizeof(*bpfEvent):]
+
+	// The DNS layer starts after the ethernet, IP and UDP headers.
+	// This is hardcoded for now as the datapath assumes the DNS packet is UDP over IPv4
+	// without any options. Next iterations will fix this.
+	// Ethernet + IP + UDP
+	const dnsOff = 14 + 20 + 8
+	dnsLayer := layers.DNS{}
+	err := dnsLayer.DecodeFromBytes(packetBytes[dnsOff:], gopacket.NilDecodeFeedback)
```

Comment:
To avoid `panic: runtime error: slice bounds out of range`:

You should check if `len(packetBytes) <= dnsOff`.

If the packet is a malformed UDP packet that has 7 bytes instead of 8 bytes, and without payload, the ebpf code currently does not check that.